executor-scripts/linux/vrrp

```diff
@@ -0,0 +1,347 @@
+# implied. In no event shall the authors be liable for any damages arising
+# from the use of this software.
+#
+# Sun, 21 Apr 2023 10:36:31 +0300
+#  -- EasyNetDev <devel@easynet.dev>
+#
+# This executor will do the configuration for MACVLAN virtual interfaces.
+#
+# Each MACVLAN interface must be controlled by a VRRP daemon, like FRRouting.
+#
+# Options needed for VRRP executro to be able to create the MACVLAN interface are:
+#
+# IFACE                Main interface name
+# IF_VRRP_ID           VRID list to create the interfaces only
+# IF_VRRP_ADDRESS      IP list to create the interfaces and add IPs (Optional)
+#
+# Interface naming would be in format vXXXXXXXX-YY-V
+# Where
+#  ->  XXXXXXXX is CRC32 value calculated for main interface name
+#  ->  YY	is HEX value of the VRRP ID
+#  ->  V	is IP version 4 or 6
+#
+# This keeps the a fix VRRP interface name length to 14 characters (maximum are 15 characters in Linux)
+# and the part vXXXXXXXX-YY is the same for both IP protocols.
+# Also it keeps the same names after reboots if the main interface name is not changed or the CRC32 tool is not changed.
+#
+# This script is written in shell and is not using any Bash built-in functions.
+#
+
+[ -n "$VERBOSE" ] && set -x
+
+# Global variables for CRC32 tool
+crc32_cmd=""
+crc32_tool=""
+
+################################################################################
+#                        VRRP CRC32 tool functions                             #
+################################################################################
+
+# In case the crc32_tool is in auto, search for crc32b then for cksum
+find_crc32_tools_in_auto() {
+    # Search for crc32b
+    # In case we are in test mode, check for current build directory for crc32b
+    if [ -n "${MOCK}" -a -z ${crc32_cmd} ]; then
+	crc32_tool="MOCK"
+	return
+    fi
+    crc32_cmd=$(which crc32sum)
+    if [ -n "${crc32_cmd}" ]; then
+	crc32_tool="crc32sum"
+	return
+    fi
+    # Search for cksum
+    crc32_cmd=$(which cksum)
+    if [ -n "${crc32_cmd}" ]; then
+	crc32_tool="cksum"
+	return
+    fi
+    # If none of them are found, print an error message
+    if [ -z "${crc32_cmd}" ]; then
+	echo "Couldn't find crc32sum or cksum tools! Please build ifupdown-ng with crc32sum support or install coreutils package!"
+	exit 0
+    fi
+}
+
+# Compute CRC32 value using the selected CRC32 tool
+compute_crc32_hex() {
+    case "$crc32_tool" in
+	crc32sum|cksum)
+	    printf "%08x" $(echo -n "$1" | ${crc32_cmd} | cut -f1 -d" ")
+	    ;;
+	MOCK)
+	    # In test mode, check if we have crc32sum compiled, cksum installed or just return CRC32 for "eth0"
+	    if [ -x "./crc32sum" ]; then
+		printf "%08x" $(echo -n "$1" | .//crc32sum | cut -f1 -d" ")
+	    elif [ -x $(which cksum) ]; then
+		printf "%08x" $(echo -n "$1" | $(which cksum) | cut -f1 -d" ")
+	    else
+		echo "43bfa0f3"
+	    fi
+	    ;;
+	*)
+	    echo "Invalid CRC32 tool: $crc32_tool! Please use crc32sum or cksum!"
+	    exit 0
+	    ;;
```

Comment:
This prevents someone from overriding the tool, for instance debugging or writing a different tool.  Perhaps even a custom tool that uses some CPU acceleration.

Should be simplified with just something akin to "if MOCK echo 43bfa0f3", "if crc32_cmd is unset error out", otherwise use the crc32_cmd as provided.

Reply:
Ok, could be a solution. I can change this part of the script.

Reply:
Removed the additional code and changed to this:
```
# Compute CRC32 value using the selected CRC32 tool
compute_crc32_hex() {
    case "$crc32_tool" in
        crc32sum|cksum)
            printf "%08x" $(echo -n "$1" | ${crc32_cmd} | cut -f1 -d" ")
            ;;
        MOCK)
            # This is CRC32 hex value calculated for "eth0" string.
            echo "43bfa0f3"
            ;;
        *)
            echo "Invalid CRC32 tool: $crc32_tool! Only allowed tools are crc32sum or cksum!"
            exit 0
            ;;
    esac
}
```